src/uu/tail/src/args.rs

```diff
@@ -242,6 +237,82 @@ impl Settings {
 
         Ok(settings)
     }
+
+    pub fn has_only_stdin(&self) -> bool {
+        self.inputs.iter().all(|input| input.is_stdin())
+    }
+
+    pub fn has_stdin(&self) -> bool {
+        self.inputs.iter().any(|input| input.is_stdin())
+    }
+
+    pub fn num_inputs(&self) -> usize {
+        self.inputs.len()
+    }
+
+    pub fn check_warnings(&self) -> CheckResult {
+        if self.retry {
+            if self.follow.is_none() {
+                show_warning!("--retry ignored; --retry is useful only when following");
+            } else if self.follow == Some(FollowMode::Descriptor) {
+                show_warning!("--retry only effective for the initial open");
+            }
+        }
+
+        let mut no_pid_support: Option<CheckResult> = None;
+        if self.pid != 0 {
+            if self.follow.is_none() {
+                show_warning!("PID ignored; --pid=PID is useful only when following");
+            } else if !platform::supports_pid_checks(self.pid) {
+                show_warning!("--pid=PID is not supported on this system");
+                no_pid_support = Some(CheckResult::NoPidSupport);
+            }
+        }
+
+        if self.follow.is_some() && !self.default_stdin && self.has_stdin() {
+            let blocking_stdin = (no_pid_support.is_some() || self.pid == 0)
+                && self.follow == Some(FollowMode::Descriptor)
+                && self.num_inputs() == 1
+                && Handle::stdin().map_or(false, |handle| {
+                    handle
+                        .as_file()
+                        .metadata()
+                        .map_or(false, |meta| !meta.is_file())
+                });
+
+            if !blocking_stdin && atty::is(Stream::Stdin) {
+                show_warning!("following standard input indefinitely is ineffective");
+            }
+        }
+
+        if let Some(check) = no_pid_support {
+            check
+        } else {
+            CheckResult::Ok
+        }
+    }
+
+    pub fn verify(&self) -> VerificationResult {
```

Comment:
I like that you refactored `verify` into a separate function. However, `VerificationResult` feels like a `UResult<()>` with extra steps though, with the disadvantage that you can use `?` on it. The mapping is basically:
- `VerificationResult::Ok` -> `Ok(())`
- `VerificationResult::Exit` -> [`ExitCode`](https://docs.rs/uucore/latest/uucore/error/struct.ExitCode.html).
- `VerificationResult::Fatal` -> `USimpleError`

Reply:
I like that. You're totally right, that way it looks like UResult, but using a UResult<Option<i32>> or alike is also a little bit ugly and obscure. Now that I look at VerificationResult it wasn't good choice of enum constants. Plain enums without value like VerificationResult::{Ok, CannotFollowByName, NoOutput} would have been better. Then the decision about what to do is moved into the calling function and it's easily extensible if needed. Would this solve your concerns?

Reply:
It'd be better I suppose, but since you don't have to do anything with the result except turn it into an error, I feel like just creating an error makes the most sense. In my opinion it's a situation where the additional type makes it more complicated instead of making it more clear. Either it's fine or we immediately exit with an error.

Reply:
Sorry, haven't seen your last comment. I've pushed a version of VerificationResult I was talking about before.

Reply:
> Either it's fine or we immediately exit with an error.

Can't do this because we also exit without error. Maybe the updated version helps a little bit.

Reply:
We can actually. `USimpleError`, `ExitCode` and friends all support exit code `0`. I admit it's a bit weird, but if you interpret `Err` as "exit the program with maybe an error" it makes sense.

> Sidenote: Once the `Try` trait stabilizes we might be able to create a custom `Result` or something that has a `exit_with_success` kinda variant, but it's not really a pressing issue.

Reply:
I don't really like that. We're exchanging something clear without hidden stuff against something obscure and weird. What's wrong with a simple and punctly enum thats coming with 0 costs? When I read Err(..) or USimpleError, I'm instantly thinking about an error and expect an error result but returning with 0 is definitely not an error.

Reply:
Must be that I'm very familiar with all the `UResult` stuff cause I wrote it then. Feel free to leave it like it is then.

Reply:
I got curious if we already did something like the error handling I suggested elsewhere. Turns out we do in exactly one spot, which actually should probably be a `show_warning!`:

https://github.com/uutils/coreutils/blob/0324a9480e7362d7436c0406331185d7b24adcad/src/uu/tr/src/tr.rs#L75-L82

So yeah, it's probably not how people tend to think about it.

> What's wrong with a simple and punctly enum thats coming with 0 costs?

There's one tiny cost, which is that the effect is decoupled from the check. Let's say I run the program and run into this error and I want to check where it came from. I then search the code for the error string and find that it's caused by some enum and now I have to find where that enum appears and I can find where my error originated. Again, not a big deal but this a tiny bit of additional abstraction. Arguably, however, one could also argue that we should go all in on the enum approach, not just for verification but all errors. Much like how other utils like `cat` or `ls` handle their errors.

Reply:
> Feel free to leave it like it is.

Ok.

> Arguably, we should go all in on the enum approach, not just for verification but all errors. 

I'm totally on your side. I like it a lot how rust made enums, types, matches etc. from functional style their own.

> ... but that's probably best left to another PR.

We could talk about it in #3952, if you like to (and when I'm ready with UChild and drawing some more commits out). I think we've then got a proper base for it there.